stdlib/toolchain/Compatibility53/BuiltinProtocolConformances.cpp

```diff
@@ -0,0 +1,136 @@
+//===--- BuiltinProtocolConformances.cpp ----------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Definitions of some builtin protocol witnesses.
+//
+//===----------------------------------------------------------------------===//
+
+#include "Overrides.h"
+#include "swift/Basic/Lazy.h"
+#include "swift/Runtime/BuiltinProtocolConformances.h"
+#include "swift/Runtime/Casting.h"
+#include <dlfcn.h>
+#include <mach-o/dyld.h>
+
+using namespace swift;
+
+static const ProtocolDescriptor *getEquatableDescriptor() {
+  auto descriptor = SWIFT_LAZY_CONSTANT(
+    reinterpret_cast<const ProtocolDescriptor *>(
+                     dlsym(RTLD_DEFAULT, "$sSQMp")));
+  return descriptor;
+}
+
+static const WitnessTable *conformsToProtocol(const Metadata *type,
+                                        const ProtocolDescriptor *protocol) {
+  using Fn = const WitnessTable *(const Metadata *, const ProtocolDescriptor *);
+  auto func = SWIFT_LAZY_CONSTANT(
+    reinterpret_cast<Fn *>(
+      dlsym(RTLD_DEFAULT, "swift_conformsToProtocol")));
+  return func(type, protocol);
+}
+
+#define TUPLE_EQUATABLE_WT SYMBOL("_swift_tupleEquatable_wt")
+
+// Define the conformance descriptor for tuple Equatable. We do this in
+// assembly to work around relative reference issues.
+__asm(
+  "  .section __DATA,__data\n"
+  "  .globl " TUPLE_EQUATABLE_CONF "\n"
+  "  .p2align 2\n"
+  TUPLE_EQUATABLE_CONF ":\n"
+  // This is an indirectable relative reference to the Equatable protocol
+  // descriptor. However, this is 0 here because the compatibility libraries
+  // can't have a dependency on libswiftCore (which is where Equatable live).
+  "  .long 0\n"
+  // 769 is the MetadataKind::Tuple
+  "  .long 769\n"
+  // This is a direct relative reference to the witness table defined below.
+  "  .long ((" TUPLE_EQUATABLE_WT ") - (" TUPLE_EQUATABLE_CONF ")) - 8\n"
+  // 32 are the ConformanceFlags with the type reference bit set to MetadataKind.
+  "  .long 32\n"
+);
+
+extern const ProtocolConformanceDescriptor _swift_tupleEquatable_conf;
+
+// Due to the fact that the compatibility libraries can't have a hard
+// dependency to libswiftCore (which is where the Equatable protocol desciptor
+// live), we have to manually implant this before calling any user code.
+__attribute__((constructor))
+void _implaceEquatableDescriptor() {
+  auto tupleEquatableConf = const_cast<int32_t *>(
+    reinterpret_cast<const int32_t *>(&_swift_tupleEquatable_conf));
```

Comment:
For extremely pedantic correctness, instead of `int32_t *`, `decltype(reinterpret_cast<ProtocolConformanceDescriptor *>(nullptr_t).Protocol)::Offset`. Which, I mean, won't work, even if my syntax was correct (which it isn't, my C++ is way out of practice), but it's gotta be possible to at least figure out the appropriate `Offset` type for `InProcess::RelativeIndirectablePointer`, I'd think...

Reply:
Or maybe better yet (emphasis on maybe), `const_cast<std::remove_const_t<reinterpret_cast<ProtocolConformanceDescriptor const *> > >(&_swift_tupleEquatable_conf)`, then `tupleEquatableConf->Protocol = ...`, which also won't work because the ivar is private, among a half dozen other reasons... What I'm getting at regardless is see if you can make the types at least a little more generic 🤷‍♀️. If it even matters.